### PR TITLE
feat: add native low precision decentralized operator

### PR DESCRIPTION
**Pull request by wangraying**

### CODE REVIEW

File: bagua-core-internal/src/datatypes/mod.rs [hunk 1]
```diff
@@ -983,6 +1072,7 @@ impl BaguaBucket {
                 tensors: tensors.iter().map(|x| (**x).clone()).collect(),
                 comm_ops: vec![],
                 dtype: tensors.first().unwrap().inner.read().raw.dtype().clone(),
+                states: Default::default(),
```
Comment (NOBLES5E) [inline bagua-core-internal/src/datatypes/mod.rs:1075]: don't do this, make operator specific logic local to the operator itself.

you can do stateful operator.

File: bagua-core-internal/src/comm_ops/decentralized_low_precision_synchronous.rs [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+use crate::datatypes::{BaguaBucket, BaguaTensorRaw, RawBaguaTensor, TensorCompressionMethod};
+use crate::comm_ops::decentralized_full_precision_synchronous::PeerSelectionMode;
+use crate::resource_pool::CUDA_DEVICE_MEMORY_POOL;
+use crate::events::BaguaEventChannel;
+use crate::{BaguaCommOpChannels, BaguaScheduledCommOp};
+use parking_lot::Mutex;
+use std::sync::Arc;
+
+#[derive(Debug)]
+pub struct DecentralizedLowPrecisionSynchronousPostStep {
```
Comment (NOBLES5E) [inline bagua-core-internal/src/comm_ops/decentralized_low_precision_synchronous.rs:12]: "post_step" etc. are planned to be deprecated to simplify operator concept

Comment (wangraying) [reply bagua-core-internal/src/comm_ops/decentralized_low_precision_synchronous.rs:12]: About the `then()` design, is it possible to share states between different operators?

File: bagua-core-internal/src/datatypes/mod.rs [hunk 3]
```diff
@@ -1128,14 +1168,16 @@ impl BaguaBucket {
                      compression_method: TensorCompressionMethod::MinMaxUInt8(
                          MinMaxUInt8CompressionParameters {},
                      ),
+                    weight: weight.expect("cannot get weight"),
+                    left_peer_weight: left_peer_weight.expect("cannot get left_peer_weight"),
+                    right_peer_weight: right_peer_weight.expect("cannot get right_peer_weight"),
```
Comment (NOBLES5E) [inline bagua-core-internal/src/datatypes/mod.rs:1173]: need more detailed error message, to help user fix

File: bagua-core-internal/src/comm_ops/decentralized_low_precision_synchronous.rs [hunk 4]
```diff
@@ -167,65 +44,143 @@ impl CommOpTrait for DecentralizedLowPrecisionSynchronous {
...
+            device_id: self.left_peer_weight.device_id(),
+            pool_allocations: vec![],
+        };
+        let mut right_peer_weight = BaguaTensorRaw {
+            ptr: self.right_peer_weight.data_ptr(),
+            num_elem: self.right_peer_weight.num_elements(),
+            dtype: self.right_peer_weight.inner.read().raw.dtype(),
+            num_elem_allocated: self.right_peer_weight.num_elements_allocated(),
+            device_id: self.right_peer_weight.device_id(),
+            pool_allocations: vec![],
```
Comment (NOBLES5E) [inline bagua-core-internal/src/comm_ops/decentralized_low_precision_synchronous.rs:69]: seems unnecessary to convert to `BaguaTensorRaw`?

File: bagua-core-py/src/lib.rs [hunk 5]
```diff
@@ -320,6 +335,19 @@ impl BaguaBucketPy {
...
+            .states()
+            .iter()
+            .map(|(k, v)| (k.clone(), BaguaTensorPy {inner: v.clone()}))
+            .collect()
+    }
+
+    pub fn set_state(&mut self, name: String, tensor: PyRef<BaguaTensorPy>) -> PyResult<()>{
+        self.inner.set_state(name, (*tensor).inner.clone());   
+        Ok(())
+    }
```
Comment (NOBLES5E) [inline bagua-core-py/src/lib.rs:349]: as last discussion, don't do this

Comment (wangraying) [reply bagua-core-py/src/lib.rs]: > as last discussion, don't do this

removed

**Pull request merged**